highstreets/features/processing_functions.py

```diff
@@ -0,0 +1,154 @@
+import pandas as pd
+from sklearn.linear_model import LinearRegression
+
+
+# create function to take averages of every x months
+# and see which hs fall in which proportion x months after that
+def create_labels(hs, label_month, comparison_start, comparison_end):
+    """Labels high streets at risk, depending on whether
+    spending is in which quartile of average of x months prior"""
+    hs_latest = hs.loc[(hs["month_year"] == label_month)]
+    hs_latest = hs_latest.groupby(["highstreet_name"]).mean()
+    # turn comp start and comp end variables into dates and find sequence btween
+    comparison_start = pd.to_datetime(comparison_start)
+    comparison_end = pd.to_datetime(comparison_end)
+    date_range = pd.date_range(comparison_start, comparison_end, freq="m").to_period(
+        "M"
+    )
+    hs_compare = hs.loc[(hs["month_year"]).isin(date_range)]
+    hs_compare = hs_compare.groupby(["highstreet_name"]).mean()
+
+    # merge dataframe
+    merged_df = pd.merge(hs_latest, hs_compare, on="highstreet_name")
+    # condtional labelling, if latest value not at least at average level, label at risk
+    merged_df["proportion"] = (
+        merged_df["txn_amt_wd_retail_x"] / merged_df["txn_amt_wd_retail_y"]
+    )
+    merged_df["labels"] = pd.qcut(
+        merged_df["txn_amt_wd_retail_x"] / merged_df["txn_amt_wd_retail_y"],
+        4,
+        labels=[0, 1, 2, 3],
+    )
+    merged_df = merged_df.reset_index()
+    # subset relevant columns
+    merged_df = merged_df[["labels"]]
+    return merged_df
+
+
+# loop and select data for relevant months
+def create_mean_sd_o2(highstreet_df, predictor_days):
+    df = []
+    for days in predictor_days:
+        df.append(highstreet_df.loc[(highstreet_df["count_date"] == days)])
+    df = pd.concat(df)
+
+    # group to create mean and std across all months
+    mean = df.groupby(["highstreet_name"]).mean().reset_index()
+    # rename and merge so less possibility of mistakes
+    mean = mean.rename({"h13": "mean_o2"}, axis=1)
+    std = df.groupby(["highstreet_name"]).std().reset_index()
+    std = std.rename({"h13": "std_o2"}, axis=1)
+    # merge and select columns
+    features = pd.merge(mean, std, on="highstreet_name")[
+        ["highstreet_name", "mean_o2", "std_o2"]
+    ]
+    return features
+
+    # loop and select data for relevant months
+
+
+def create_mean_sd_mcard(highstreet_df, predictor_months):
+    df = []
+    for months in predictor_months:
+        df.append(highstreet_df.loc[(highstreet_df["month_year"] == months)])
+    df = pd.concat(df)
+
+    # group to create mean and std across all months
+    mean = df.groupby(["highstreet_name"]).mean().reset_index()
+    # rename and merge so less possibility of mistakes
+    mean = mean.rename({"txn_amt_wd_retail": "mean"}, axis=1)
+    std = df.groupby(["highstreet_name"]).std().reset_index()
```

Comment:
Small thing but you could consider making a groupby object and then extracting from that object everything you need (mean and std here) instead of doing the groupby repeatedly.